pysos.py

```diff
@@ -394,35 +359,41 @@ def append(self, value):
         self._indexes.append(key)
         
     def __delitem__(self, i):
-        # trigger observers
-        if self._observers:
-            old_value = self[i]
-            for callback in self._observers:
-                callback(i, None, old_value)
-                
+        self._trigger_observers(i, None, self[i])
         key = self._indexes[i]
         del self._dict[key]
         del self._indexes[i]
     
     def __len__(self):
         return len(self._indexes)
         
-    def __contains__(self, key):
-        raise Exception('Operation not supported for lists')
-    
+    def __contains__(self, value):
```

Comment:
I would add a comment that this operation is vers costly because it triggers a whole file scan.

Reply:
I also was thinking about that. It is your decision. Since on loading time pysos reads all the file I think it is not for dicts and dictionaries which are too big to fit in memory. Reading the whole line means that every item was in the memory at least once. This steps has only one addition: they are json parsed now. I think it is not a big difference.

I would like to mention that implementing the slice (`my_list[3:8]`) reads everything into memory.

Again, it is your decision, what whould you like me to do?

Reply:
Here is a test file:

```python
import pathlib
import pickle
import sys
from contextlib import contextmanager
from time import perf_counter

import pysos

pysos_file = pathlib.Path("temp/big-list.sos")
pickle_file = pathlib.Path("temp/big-list.pickle")

NUMBER_OF_ITEMS = 1_000_000


@contextmanager
def measure_time(name):
    print(name, end=": ")
    sys.stdout.flush()
    start = perf_counter()
    yield
    elapsed = perf_counter() - start
    print(f"{elapsed:.2f} secs")


def value_for_index(index):
    return f"This is the value number {index}. " * 100


native_list = []

with measure_time("Create a big list fit to the memory"):
    for index in range(NUMBER_OF_ITEMS):
        native_list.append(value_for_index(index))

with measure_time("Persist native list with pickle"):
    with pickle_file.open("wb") as f:
        pickle.dump(file=f, obj=native_list)

with measure_time("Load persisted native list with pickle"):
    with pickle_file.open("rb") as f:
        native_list = pickle.load(file=f)

file_size = pickle_file.stat().st_size / 1024 / 1024 / 1024
print(f"Pickle file size: {file_size:.1f} Gb")

del native_list

with measure_time("Add items to persisted list"):
    persisted_list = pysos.List(pysos_file)
    persisted_list.clear()
    for index in range(NUMBER_OF_ITEMS):
        persisted_list.append(value_for_index(index))
    del persisted_list

file_size = pysos_file.stat().st_size / 1024 / 1024 / 1024
print(f"Pysos file size: {file_size:.1f} Gb")

with measure_time("Load the list"):
    persisted_list = pysos.List(pysos_file)

with measure_time("Lookup for a value in the middle"):
    persisted_list.index(value_for_index(NUMBER_OF_ITEMS // 2))
```

The results:

```
Create a big list fit to the memory: 2.00 secs
Persist native list with pickle: 4.24 secs
Load persisted native list with pickle: 3.72 secs
Pickle file size: 3.1 Gb
Add items to persisted list: 50.34 secs
Pysos file size: 3.1 Gb
Load the list: 14.12 secs
Lookup for a value in the middle: 8.70 secs
```

Which means that for a ~3Gb list (I have 32Gb memory), just writing `pysos.List(filename)` without any valuable action required 14 seconds. I think for such a big datastructures (which still fits to the memory) there are better solutions. We can load everything into the memory using Pickle, or use a tool which touches the value only on accessing it.

Also, who wants to look up for a big value in a huge list? They also needs to find a better data structure.

Don't be offended, I think pysos is an excellent module for its use cases.